pkg/crosscluster/logical/replication_statements.go

```diff
@@ -0,0 +1,185 @@
+			return nil, err
+		}
+	}
+
+	parameterValues := &tree.ValuesClause{
+		Rows: []tree.Exprs{
+			parameters,
+		},
+	}
+
+	rows := &tree.Select{
+		Select: parameterValues,
+	}
+
+	insert := &tree.Insert{
+		Table: &tree.TableRef{
+			TableID: int64(table.GetID()),
+			As:      tree.AliasClause{Alias: "replication_target"},
+		},
+		Rows:      rows,
+		Columns:   columnNames,
+		Returning: tree.AbsentReturningClause,
+	}
+
+	return insert, nil
+}
+
+// newUpdateStatement returns a statement that can be used to update a row in
+// the table. If a table has `n` columns, the statement will have `2n`
+// parameters, where the first `n` parameters are the previous values of the row
+// and the last `n` parameters are the new values of the row.
+//
+// Parameters are ordered by column ID.
+func newUpdateStatement(table catalog.TableDescriptor) (tree.Statement, error) {
+	columns := getPhysicalColumns(table)
+
+	// Create WHERE clause placeholders for every column in the table
+	var whereClause tree.Expr
+	for i, col := range columns {
+		placeholder, err := tree.NewPlaceholder(fmt.Sprintf("%d", i+1))
+		if err != nil {
+			return nil, err
+		}
+		colExpr := &tree.ComparisonExpr{
+			Operator: treecmp.MakeComparisonOperator(treecmp.EQ),
+			Left:     &tree.ColumnItem{ColumnName: tree.Name(col.GetName())},
+			Right:    placeholder,
+		}
+
+		if whereClause == nil {
+			whereClause = colExpr
+		} else {
+			whereClause = &tree.AndExpr{
+				Left:  whereClause,
+				Right: colExpr,
+			}
+		}
+	}
+
+	exprs := make(tree.UpdateExprs, len(columns))
+	for i, col := range columns {
+		nameNode := tree.Name(col.GetName())
+		names := tree.NameList{nameNode}
+
+		// Create a placeholder for the new value (num_columns+i+1) since we
+		// placholders to the where and placholders are 1 indexed.
+		placeholder, err := tree.NewPlaceholder(fmt.Sprintf("%d", len(columns)+i+1))
+		if err != nil {
+			return nil, err
+		}
+
+		exprs[i] = &tree.UpdateExpr{
+			Names: names,
+			Expr: &tree.CastExpr{
+				Expr:       placeholder,
+				Type:       col.GetType(),
+				SyntaxMode: tree.CastPrepend,
+			},
+		}
+	}
+
+	// Create the final update statement
+	update := &tree.Update{
+		Table: &tree.TableRef{
+			TableID: int64(table.GetID()),
+			As:      tree.AliasClause{Alias: "replication_target"},
+		},
+		Exprs:     exprs,
+		Where:     &tree.Where{Type: tree.AstWhere, Expr: whereClause},
+		Returning: tree.AbsentReturningClause,
+	}
+
+	return update, nil
+}
+
+// newDeleteStatement returns a statement that can be used to delete a row from
+// the table. The statement will have `n` parameters, where `n` is the number of
+// columns in the table. Parameters are used in the WHERE clause to precisely
+// identify the row to delete.
+//
+// Parameters are ordered by column ID.
+func newDeleteStatement(table catalog.TableDescriptor) (tree.Statement, error) {
+	columns := getPhysicalColumns(table)
+
+	// Create WHERE clause placeholders for every column in the table
```

Comment:
nit: might be able to factor this out of both here and `newUpdateStatement`

Reply:
Done!